radix-engine/src/engine/process.rs

```diff
@@ -51,38 +51,103 @@ macro_rules! re_warn {
     };
 }
 
-enum LazyMapState {
-    Uncommitted { root: Mid },
-    Committed { component_address: Address },
+/// Represents an interpreter instance.
+pub struct Interpreter {
+    invocation: Invocation,
+    module: ModuleRef,
+    memory: MemoryRef,
 }
 
-enum ComponentState {
-    Empty,
-    Loaded {
+/// Keeps invocation information.
+#[derive(Debug, Clone)]
+pub struct Invocation {
+    actor: Actor,
+    package_address: Address,
+    export_name: String,
+    function: String,
+    args: Vec<ValidatedData>,
+}
+
+/// Qualitative states for a WASM process
+enum InterpreterState {
+    Blueprint,
+    ComponentEmpty {
+        component_address: Address,
+    },
+    ComponentLoaded {
         component_address: Address,
-        component_data: ValidatedData,
+        initial_loaded_object_refs: ComponentObjectsSetRef,
+        additional_object_refs: ComponentObjectsSetRef,
```

Comment:
I see the direction you're moving towards. Component objects are grouped into a component, but access pattern must follow the reference chain. 

But I think this might deserve a proper design documentation before implementation, as this introduces a slightly more complex system. It would cover 
- How objects **within a component** are stored and accessed (complexity for read and write - which impacts metering)?
- How the reference structure is inferred?
- What assumptions have been made in user space (what if it's not Rust)?